clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,246 @@
+      (if dcReadDataCount then rCnt else deepErrorX "Read data count disabled")
+      (deepErrorX "No sample" :- rData)
+
+ where
+  rstSignalR = unsafeToHighPolarity rst
+  rstSignalW = unsafeSynchronizer rClk wClk $ unsafeToHighPolarity rst
+
+  -- reified depth
+  maxDepth = snatToNum @Int (powSNat (SNat @2) (SNat @depth))
+
+  -- TODO: how do we ensure it doesn't "space leak" in Haskell?
+  --
+  -- force slower first?
+  -- https://github.com/clash-lang/clash-compiler/blob/ea114d8edd6a110f72d148203b9db2454cae8f37/clash-prelude/src/Clash/Explicit/BlockRam.hs#L1276-L1280
+
+  go ::
+    FifoState n ->
+    Signal read Bool -> -- reset
+    Signal read Bool -> -- read enabled
+    Signal write Bool -> -- reset
+    Signal write (Maybe (BitVector n)) -> -- write data
+    ( Signal write ResetBusy
+    , Signal write Full
+    , Signal write Bool
+    , Signal write (DataCount depth)
+
+    , Signal read ResetBusy
+    , Signal read Empty
+    , Signal read Bool
+    , Signal read (DataCount depth)
+    , Signal read (BitVector n)
+    )
+  go st@(FifoState _ rt) rstR rEna rstW =
+    if rt < tWr
+      then goRead st rstR rEna rstW
+      else goWrite st rstR rEna rstW
+    -- TODO: goBoth case?
+
+  goWrite (FifoState _ rt) rstR rEna (True :- rstWNext) (_ :- wData) =
+      (True :- wRstBusy, False :- preFull, False :- preOver, 0 :- preWCnt, rRstBusy, fifoEmpty, under, rCnt, rData)
+    where
+      (wRstBusy, preFull, preOver, preWCnt, rRstBusy, fifoEmpty, under, rCnt, rData) =
+        go (FifoState mempty (rt-tWr)) rstR rEna rstWNext wData
+
+  goWrite (FifoState q rt) rstR rEna (_ :- rstW) (wDat :- wDats1) =
+    (False :- wRstBusy, full, over, wCnt, rRstBusy, fifoEmpty, under, rCnt, rData)
+    where
+      (wRstBusy, preFull, preOver, preWCnt, rRstBusy, fifoEmpty, under, rCnt, rData) =
+        go (FifoState q' (rt-tWr)) rstR rEna rstW wDats1
+
+      wCnt = sDepth q :- preWCnt
+      full = (Seq.length q == maxDepth) :- preFull
+      (q', over) =
+        if Seq.length q < maxDepth
+          then (case wDat of { Just x -> x Seq.<| q ; _ -> q }, False :- preOver)
+          else (q, True :- preOver)
+
+  sDepth = fromIntegral . Seq.length
+
+  goRead (FifoState _ rt) (True :- rstRNext) (_ :- rEnas1) rstW wData =
+    (wRstBusy, full, over, wCnt, True :- rRstBusy, fifoEmpty, under, rCnt, rData)
+    where
+      rData = deepErrorX "Reset" :- preRData
+      fifoEmpty = True :- preEmpty
+      rCnt = 0 :- preRCnt
+      under = False :- preUnder
+
+      (wRstBusy, full, over, wCnt, rRstBusy, preEmpty, preUnder, preRCnt, preRData) =
+        go (FifoState mempty (rt+tR)) rstRNext rEnas1 rstW wData
+
+  goRead (FifoState q rt) (_ :- rstRNext) (rEna :- rEnas1) rstW wData =
+    (wRstBusy, full, over, wCnt, False :- rRstBusy, fifoEmpty, under, rCnt, rData)
+    where
+      rCnt = sDepth q :- preRCnt
+      fifoEmpty = (Seq.length q == 0) :- preEmpty
+      rData = nextData :- preRData
```

Comment:
You can move this up to the `goRead` line. If you need the lazy behavior use `~(wibble :- wobble)`. This will probably help you get a consistent naming scheme for the element / signals too.

Reply:
Sure!

Reply:
Wait is that really better? I just split it out into `where` clauses since the line was long.

Reply:
Right, actually, I misread. You're constructing a signal here, not deconstructing it :see_no_evil:. Sorry.